httpcache.go

```diff
@@ -51,7 +64,145 @@ func CachedResponse(c Cache, req *http.Request) (resp *http.Response, err error)
+	} else {
+		start, err = strconv.ParseInt(strStart, 10, 64)
+		if err != nil {
+			return -1, -1, err
+		}
+		end, err = strconv.ParseInt(strEnd, 10, 64)
+		if err != nil {
+			return -1, -1, err
+		}
+	}
+	if start >= end {
+		return -1, -1, fmt.Errorf("invalid start %d >= end %d!", start, end)
+	}
+	return start, end, nil
+}
+
+// validateRanges checks that a cached request for a given response is within data that has been already loaded
+func validateRanges(start, end int64, resp *http.Response) (ok bool) {
+	// if the response cites partial content we need to compare the partial content we have stored
+	// with the ranges we require and, if not compatbile, fetch it again
+	// http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html section 14.16
+	if resp.StatusCode == http.StatusPartialContent {
+		rawContentRange := resp.Header.Get("content-range")
+		if rawContentRange == "" {
+			logger.Printf("request for %s is stored as 206-partial-content but has no content-range header!", resp.Request.URL.String())
+			return false
+		}
+		if !strings.Contains(rawContentRange, "bytes") {
+			logger.Printf("non-satisfiable range type in %s", rawContentRange)
+			return false
+		}
+		// the format is START-END/TOTAL or START-END/* if TOTAL is unknown
+		// if we find * we re-fetch the request as most probably the content was ephemereal
+		// or is highly probable iot has changed
+		if strings.Contains(rawContentRange, "*") {
+			return false
+		}
+		re := regexp.MustCompile("bytes ([0-9]+)-([0-9]+)/([0-9]+)")
+		// the first element is always the full match, skip it
+		matchedValues := re.FindStringSubmatch(rawContentRange)[1:]
+		currentStart, err := strconv.ParseInt(matchedValues[0], 10, 64)
+		if err != nil {
+			logger.Printf("cached response has malformed content-range header %s", rawContentRange)
+			return false
+		}
+		currentEnd, err := strconv.ParseInt(matchedValues[1], 10, 64)
+		if err != nil {
+			logger.Printf("cached response has malformed content-range header %s", rawContentRange)
+			return false
+		}
+		total, err := strconv.ParseInt(matchedValues[2], 10, 64)
+		if err != nil {
+			logger.Printf("cached response has malformed content-range header %s", rawContentRange)
+			return false
+		}
+		// validate the request ranges against the response headers
+		if start < currentStart || end > currentEnd || end > total {
+			logger.Printf("start: %d, currentStart: %d, end: %d, currentEnd: %d, total: %d", start, currentStart, end, currentEnd, total)
+			return false
+		}
+		return true
+		// the response is full content, use the content-length header to verify ranges
+	} else {
+		contentLength, err := strconv.ParseInt(resp.Header.Get("content-length"), 10, 64)
+		if err != nil {
+			logger.Printf("stored response has malformed or invalid content length %s", contentLength)
+			return false
+		}
+		if end > contentLength {
+			return false
+		}
+		return true
+	}
+	logger.Print("we should never ever reach this line")
```

Comment:
Instead of logging this, remove the else block from the if-statement and make its statements be the end of this function.